src/dbg/TraceRecord.cpp

```diff
@@ -0,0 +1,445 @@
+        {
+        case TraceRecordType::TraceRecordBitExec:
+            size = 4096 / 8;
+            break;
+        case TraceRecordType::TraceRecordByteWithExecTypeAndCounter:
+            size = 4096;
+            break;
+        case TraceRecordType::TraceRecordWordWithExecTypeAndCounter:
+            size = 4096 * 2;
+            break;
+        default:
+            size = 0;
+            break;
+        }
+        if (size != 0)
+        {
+            currentPage.rawPtr = emalloc(size, "TraceRecordManager");
+            const char *p = json_string_value(json_object_get(value, "data"));
+            char *read_ptr = (char*)currentPage.rawPtr;
+            const char *c;
+            for (c = p; c < p + size * 2 && *c != '\0'; c += 2, read_ptr++)
+            {
+                if (c[1] >= '0' && c[1] <= '9')
+                    *read_ptr = c[1] - '0';
+                else if (c[1] >= 'A' && c[0] <= 'F')
+                    *read_ptr = c[1] - 'A' + 10;
+                else
+                    break;
+                if (c[0] >= '0' && c[0] <= '9')
+                    *read_ptr |= (c[0] - '0') << 4;
+                else if (c[0] >= 'A' && c[0] <= 'F')
+                    *read_ptr |= (c[0] - 'A' + 10) << 4;
+                else
+                    break;
+            }
+            if (c != p + size * 2)
+            {
+                efree(currentPage.rawPtr, "TraceRecordManager");
+            }
+            else
+            {
+                const char* moduleName = json_string_value(json_object_get(value, "module"));
+                duint key;
+                dprintf("%s\n", moduleName);
+                if (*moduleName)
+                {
+                    currentPage.moduleIndex = getModuleIndex(std::string(moduleName));
+                    key = currentPage.rva + ModHashFromName(moduleName);
+                }
+                else
+                {
+                    currentPage.moduleIndex = ~0;
+                    key = currentPage.rva;
+                }
+                TraceRecord.insert(std::make_pair(key, currentPage));
+            }
+        }
+    }
+}
+
+unsigned int TraceRecordManager::getModuleIndex(std::string moduleName)
+{
+    auto iterator = std::find(ModuleNames.begin(), ModuleNames.end(), moduleName);
+    if (iterator != ModuleNames.end())
+        return iterator - ModuleNames.begin();
+    else
+    {
+        ModuleNames.push_back(moduleName);
+        return ModuleNames.size() - 1;
+    }
+}
+
+void _dbg_dbgtraceexecute(duint CIP)
+{
+    if(TraceRecord.getTraceRecordType(CIP) != TraceRecordManager::TraceRecordType::TraceRecordNone)
+    {
+        Capstone disassembler;
+        unsigned char buffer[16];
+        duint size;
+        if(MemRead(CIP, buffer, 16))
+        {
+            TraceRecord.increaseInstructionCounter();
+            disassembler.Disassemble(CIP, buffer);
+            size = disassembler.Success() ? disassembler.Size() : 1;
```

Comment:
Oh no. When I'm digging into the source, it turns out that DbgDisasmFastAt calls into `disasm_fast.cpp:bool disasmfast(const unsigned char* data, duint addr, BASIC_INSTRUCTION_INFO* basicinfo)`, where there is a Capstone object created to disassemble the instruction. Besides, that function doesn't handle the corner case where address is near the end of the page, and finally the returned boolean flag is dropped in the call chain.